### DISCUSSION TITLE
Debugging White- and Blacklists and applying of those

### DISCUSSION THREAD

**Discussion by gitsteff** [netblue30/firejail#5263] [General]
I have this in my freemind.local:
`blacklist ${HOME}`
`whitelist ${HOME}/MindMaps/`
'whitelist ${HOME}/Software/freemind/FreeMind-v1.1.0_Beta2/`
When I debug this I can see that whitelisting is applied but later on blacklisting is overwriting it.

`Whitelisting /home/x/MindMaps`
`Whitelisting /home/x/Software/freemind/FreeMind-v1.1.0_Beta2`
`Whitelisting /var/lib/dbus`
`Whitelisting /var/cache/fontconfig`
`Whitelisting /var/tmp`
`Created symbolic link /var/run -> /run`
`Created symbolic link /var/lock -> /run/lock`
`Whitelisting /tmp/.X11-unix`
`Disable /home/x`
Is this default behavior or am I able to manipulate it somehow?

Btw.: using noblacklist is not working or at least not showing up when debugging.

**Comment by gitsteff** [comment]
In the end I want just have the app started in /home/x/Software/freemind/FreeMind-v1.1.0_Beta2/ and give it access only to /home/x/MindMaps where the mindmaps are stored.

**Reply by rusty-snake** [reply]
This is whitelisting.

This is not possible because you need the software itself (`~/Software/freemind`) as well as some dotfiles like `~/.freemind` for the program to work correctly/at all.

**Reply by gitsteff** [reply]
In my imagination I could whitelist those files/dirs the app needs. I can see those access validations perfectly when the app comes up.

**Reply by rusty-snake** [reply]
>  and give it access **only** to /home/x/MindMaps where the mindmaps are stored.

**Comment by rusty-snake** [comment]
Blacklisting and Whitelisting are two independent file system restrictions "technologies" in firejail.

**Comment by rusty-snake** [comment]
> using noblacklist is not working

What did you tried?

> **Note**: A `noblacklist` must match the path of the `blacklist` (after globbing is expanded).  i.e. to `noblacklist` `blacklist ${HOME}` you must write `noblacklist ${HOME}` and not `noblacklist ${HOME}/Documents`.

**Comment by ghost** [comment]
Hello, I have a similar issue, I'm trying to `blacklist` a sudirectory of a directory I did `noblacklist` but the `blacklist` directive isn't applied.
How could I circumvent this?

**Reply by rusty-snake** [reply]
Can you post the configs.

First check that

1. There is no `noblacklist` for it.
2. The file with the `blacklist` is included.
3. There is no typo in the path.

**Reply by ghost** [reply]
It's not on the same profile and application, but I was asking if it is possible to blacklist a directory inside a directory I have set noblacklist.
If it is not possible, is there a workaround?

**Reply by rusty-snake** [reply]
It is possible. If not that's a bug. But more likely a configuration or testing issue.

**Comment by ghost** [comment]
like:
```
A
|--> B
     |--> C
     |--> D
```
if `B` is blacklisted in some common `.inc` profiles inherited, I am using `noblacklist` on it.
Then I do not want the program to access to `D`, then  I am using `blacklist` but it does not have any effect.

**Reply by rusty-snake** [reply]
```console
$ mkdir -p A/B/{C,D}
$ echo 'blacklist ${HOME}/A/B' > b.inc
$ echo 'noblacklist ${HOME}/A/B' > a.profile
$ echo 'blacklist ${HOME}/A/B/D' >> a.profile
$ echo 'include b.inc' >> a.profile
$ firejail --profile=a.profile ls -R A
Reading profile a.profile
Reading profile b.inc
A:
B

A/B:
C  D

A/B/C:
ls: cannot open directory 'A/B/D': Permission denied
```

Can you give better STR. Did you tried with ABCD or with actual paths? If so which paths? How did you ordered the commands? Was there a typo (double check)? How did you checked?

**Reply by ghost** [reply]
Lines 3 and 4 in your example are in a.local instead of a.profile.
Yes, it us just an example path.
I am using another path but the behavior in this example is the same.

**Reply by rusty-snake** [reply]
>  I am using another path but the behavior in this example is the same.

What? You said you "do not want the program to access to D" and that's what happens in my example, ls tries to access D and gets "Permission denied".

**Reply by ghost** [reply]
Yes, path tree is the same but a.local is loaded begore the rest on a.profile
i think the order here can affect the behavior

**Reply by ghost** [reply]
That means lines 3 and 4 are read before line 2
